nomad/core_sched_test.go

```diff
@@ -467,44 +511,174 @@ func TestCoreScheduler_EvalGC_Batch_OldVersion(t *testing.T) {
 	core := NewCoreScheduler(s1, snap)
 
 	// Attempt the GC
-	gc := s1.coreJobEval(structs.CoreJobEvalGC, 2000)
+	gc := s1.coreJobEval(structs.CoreJobEvalGC, jobModifyIdx)
+	err = core.Process(gc)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// All allocations should remain in place.
+	assertCorrectEvalAlloc(
+		memdb.NewWatchSet(),
+		eval,
+		[]*structs.Allocation{alloc, alloc2, alloc3, alloc4},
+		[]*structs.Allocation{},
+	)
+
+	// Attempt the GC while moving the time forward significantly so that our threshold is
+	// breached.
+	tt := s1.fsm.TimeTable()
+	tt.Witness(2*jobModifyIdx, time.Now().UTC().Add(-1*s1.config.BatchEvalGCThreshold))
+
+	gc = s1.coreJobEval(structs.CoreJobEvalGC, jobModifyIdx*2)
 	err = core.Process(gc)
 	if err != nil {
 		t.Fatalf("err: %v", err)
 	}
 
-	// Alloc1 and 2 should be there, and alloc3 should be gone
+	// The only remaining allocations are those which are "current", or such that
+	// the batch job is still using them. To spell it out:
+	//
+	// Alloc1 and Alloc 2 should remain due to the update threshold.
+	// Alloc 3 should be GCed due to age
+	// Alloc 4 should remain due to non-terminal state
+	assertCorrectEvalAlloc(
+		memdb.NewWatchSet(),
+		eval,
+		[]*structs.Allocation{alloc, alloc2, alloc4},
+		[]*structs.Allocation{alloc3},
+	)
+
+	// The job should still exist.
+	ws := memdb.NewWatchSet()
+	outB, err := store.JobByID(ws, job.Namespace, job.ID)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if outB == nil {
+		t.Fatalf("bad: %v", outB)
+	}
+}
+
+// An EvalGC should reap allocations from jobs with a newer modify index and reap the eval itself
+// if all allocs are reaped.
+func TestCoreScheduler_EvalGC_Batch_OldVersionReapsEval(t *testing.T) {
```

Comment:
These tests have a _lot_ of setup code (as I'm sure you noticed!) and the assertion is only subtly different than the previous one. It might make the whole test more understandable if we collapsed these into a single test and had this one be the final assertion that the eval is GC'd once the conditions are right.